### DISCUSSION TITLE
Adding Go programming language support to Astronomy Engine

### DISCUSSION THREAD

**Discussion by cosinekitty** [cosinekitty/astronomy#320] [General]
Hi @DannyRavi, I'm starting this discussion page so we can keep track of our work to add Go language support.

I created a [spreadsheet of function dependences](https://docs.google.com/spreadsheets/d/1R2hW9h36A_XCjaFHYZ6KtIHNX274GD_T_dTKQqQ52C4/edit?usp=sharing) based on all the functions in the C# version, and how they call each other. It shows the order that we can define the functions. Some functions are private, others are public. You can tell the difference by referring to the C# source file [astronomy.cs](https://github.com/cosinekitty/astronomy/blob/master/source/csharp/astronomy.cs).

Each row of the spreadsheet has a "parent" column, which is another function to be implemented, followed by a list of all the child functions that must be implemented first.

The data and code for making this spreadsheet are in the [dependencies](https://github.com/cosinekitty/astronomy/tree/golang/dependencies) directory.

Before you do anything, I want to get started with one or two functions and show you all the stuff that needs to happen each time. This will include documentation and unit tests for each public function.

**Comment by DannyRavi** [comment]
Hi @cosinekitty Thank you for making these documentaries. I am waiting for you to do the part related to those two functions and then I will continue.
Also, in order to better understand the structure of the program, I want to implement the C code of this program completely on Arduino (esp8266/esp32).

**Comment by cosinekitty** [comment]
> I am waiting for you to do the part related to those two functions and then I will continue.

OK, that sounds good. It will take me a little while longer as I work through the initial issues of generating online documentation. I have started changing the comments in the code to reflect the [Go Doc Comments](https://go.dev/doc/comment) format. I'm also looking at the [gomarkdoc](https://github.com/princjef/gomarkdoc) utility to create the markdown format documentation to be hosted right here on GitHub.

> I want to implement the C code of this program completely on Arduino (esp8266/esp32).

That sounds very interesting! I will be eager to find out how that goes. I also like to experiment with Arduino, Raspberry Pi, etc.

**Reply by DannyRavi** [reply]
Hi @cosinekitty. 
I spent some time on the implementation of the astronomy engine library on Arduino and it seems that the program can be run on hardware that has suitable hardware resources with a few changes. Also, the information about the consumed resources is clear below 
![image](https://github.com/cosinekitty/astronomy/assets/4340829/70a2cf83-8858-48ee-9db9-ea981084c1e3)
 and the link to the relevant codes is [here](https://drive.google.com/file/d/1v7eBsHCZj0OpnXMrWcO6_qMnyeW_eb50/view?usp=drive_link), and it is better to install Arduino IDE from [here](https://www.arduino.cc/en/software) and add esp8266 library to it like this [link](https://www.instructables.com/Steps-to-Setup-Arduino-IDE-for-NODEMCU-ESP8266-WiF/) ,also this work is possible without the need of an electronic board. Of course there are some problems, for example, in Arduino, the real time of the system cannot be obtained without the help of external tools such as RTC clock, GPS, and serial-port and some other solutions,another problem is the argument cannot be passed to the program except by the serial-port. In the ecliptic_vector application , we have a problem with the operating system (freeRTOS)[https://en.wikipedia.org/wiki/FreeRTOS] on esp8266, because the program task probably takes too long, and the RTOS kill this task and part of its memory is thrown out, the image this problem available below.
![ardoinoExport](https://github.com/cosinekitty/astronomy/assets/4340829/22a6062e-2da7-401e-8f64-6421b8a3d5f5)

Finally, the output of the two programs on PC and Arduino is almost same.
I suggest that we have a new branch for Arduino and work on this issue as well.

**Reply by cosinekitty** [reply]
This is very interesting! I am glad you want to work on this. I have created an `arduino` branch like you suggested. Feel free to create a new directory `demo/arduino` where you can put as many Arduino sketches as you want. I think we should keep them separate from the `demo/c` directory because the environment is more specialized.

We can also consider adding preprocessor definitions to allow people to conditionally compile the C version of Astronomy Engine so that they can remove large parts of code they don't need. This will probably help a lot of people who want to fit a subset of Astronomy Engine into a smaller microcontroller.
